src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -19,9 +19,33 @@ internal static class DeserializationHelpers
     {
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
-        { 
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
-                       
+        {
+            IEdmProperty edmProperty;
+
+            // Extract edmProperty using a Case-Sensitive match
+            edmProperty = resourceType.FindProperty(property.Name);
+
+            if (readContext != null && !readContext.DisableCaseInsensitiveRequestPropertyBinding && edmProperty == null)
+            {
+                // if edmProperty is null, we try a Case-Insensitive match.
+                bool propertyMatched = false;
+
+                IEnumerable<IEdmStructuralProperty> sProperties = resourceType.StructuralProperties();
+                foreach(IEdmStructuralProperty sProperty in sProperties)
+                {
+                    if(sProperty.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // We throw an exception when we have more than 1 case-insensitive matches and no case sensitive matches.
+                        if (propertyMatched)
+                        {
+                            throw new ODataException(Error.Format(SRResources.CannotDeserializeUnknownProperty, property.Name, resourceType.Definition));
+                        }
+                        propertyMatched = true;
+                        edmProperty = resourceType.FindProperty(sProperty.Name);
+                    }
```

Comment:
@KenitoInc Consider if this might be easier to follow than the `foreach` loop:
```csharp
IList<IEdmStructuralProperty> matchedProperties = structuralProperties.Where(d => d.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase)).ToList();

if (matchedProperties.Count > 1)
{
	// We throw an exception when we have more than 1 case-insensitive
	throw new ODataException(Error.Format(SRResources.CannotDeserializeUnknownProperty, property.Name, resourceType.Definition));
}
else if (matchedProperties == 1)
{
	edmProperty = resourceType.FindProperty(matchedProperties.First().Name);
}
```

Reply:
Calling Count against an IQueryable is expensive.